backend/src/layer/logic/api_command/command_handler.rs

```diff
@@ -0,0 +1,211 @@
+        api_command::{AuthInfo, Command, CommandError, Result},
+        image_hoster::ImageHoster,
+        persistent_data::CommandDataAccess,
+    },
+    layer::logic::api_command::auth::{
+        authenticator::Authenticator, image_command_type::ImageCommandType,
+    },
+    util::{Date, ReportReason, Uuid},
+};
+
+const REPORT_FACTOR: f64 = 1.0 / 35.0;
+
+pub struct CommandHandler<D, A, I>
+where
+    D: CommandDataAccess + Sync + Send,
+    A: AdminNotification + Sync + Send,
+    I: ImageHoster + Sync + Send,
+{
+    command_data: D,
+    admin_notification: A,
+    image_hoster: I,
+    auth: Authenticator,
+}
+
+impl<D, A, I> CommandHandler<D, A, I>
+where
+    D: CommandDataAccess + Sync + Send,
+    A: AdminNotification + Sync + Send,
+    I: ImageHoster + Sync + Send,
+{
+    /// A function that creates a new [`CommandHandler`]
+    ///
+    /// # Errors
+    /// Throws an error, if the api keys could not be gotten from [`command_data`]
+    pub async fn new(command_data: D, admin_notification: A, image_hoster: I) -> Result<Self> {
+        let keys: Vec<String> = command_data
+            .get_api_keys()
+            .await?
+            .into_iter()
+            .map(|x| x.key)
+            .collect();
+        Ok(Self {
+            command_data,
+            admin_notification,
+            image_hoster,
+            auth: Authenticator::new(keys),
+        })
+    }
+
+    fn will_be_hidden(date: Date, report_count: u32) -> bool {
+        Self::get_date_difference(date) < 30 && report_count > Self::get_report_barrier(date)
+    }
+
+    fn get_date_difference(date: Date) -> i64 {
+        let today = Local::now().date_naive();
+        let difference = today.signed_duration_since(date);
+        difference.num_days()
+    }
+
+    fn get_report_barrier(date: Date) -> u32 {
+        let t = Self::get_date_difference(date) as f64;
+        (REPORT_FACTOR * t * t).floor() as u32 + 5
+    }
+}
+
+#[async_trait]
+impl<D, A, I> Command for CommandHandler<D, A, I>
+where
+    D: CommandDataAccess + Sync + Send,
+    A: AdminNotification + Sync + Send,
+    I: ImageHoster + Sync + Send,
+{
+    /// Command to report an image. It als gets checked whether the image shall get hidden.
+    async fn report_image(
+        &self,
+        image_id: Uuid,
+        reason: ReportReason,
+        auth_info: AuthInfo,
+    ) -> Result<()> {
+        if let Some(auth_info) = auth_info {
+            let image_command_type = ImageCommandType::ReportImage(reason);
+            self.auth
+                .authn_image_command(&auth_info, image_id, image_command_type)?;
+            let info = self.command_data.get_image_info(image_id).await?;
+            if !info.approved {
+                self.command_data
+                    .add_report(image_id, auth_info.client_id, reason)
+                    .await?;
+                let image_got_hidden = Self::will_be_hidden(info.upload_date, info.report_count);
```

Comment:
should that not rather be called `should_hide` as the image did not already get hidden?